In [1]:
import os
import json
import pandas as pd
import numpy as np
import pickle
from collections import defaultdict

import os

# Set paths untuk local dataset
base_path = os.path.abspath('.')
cva_path = os.path.join(base_path, 'CVA')
hs_path = os.path.join(base_path, 'HS')

print(f"Base path: {base_path}")
print(f"CVA path: {cva_path}")
print(f"HS path: {hs_path}")
print(f"CVA folder exists: {os.path.exists(cva_path)}")
print(f"HS folder exists: {os.path.exists(hs_path)}")

Base path: c:\Users\fizel\OneDrive\Documents\Proyek Penelitian_Progress
CVA path: c:\Users\fizel\OneDrive\Documents\Proyek Penelitian_Progress\CVA
HS path: c:\Users\fizel\OneDrive\Documents\Proyek Penelitian_Progress\HS
CVA folder exists: True
HS folder exists: True


In [2]:
# Folder lokal sudah tersedia, tidak perlu extract
# Dataset sudah ada di folder CVA dan HS

# Verifikasi struktur folder
print("=== CVA Folders ===")
if os.path.exists(cva_path):
    cva_folders = sorted([d for d in os.listdir(cva_path) if os.path.isdir(os.path.join(cva_path, d))])
    print(f"Total CVA folders: {len(cva_folders)}")
    print(f"First few: {cva_folders[:5]}")
else:
    print("CVA folder tidak ditemukan!")

print("\n=== HS Folders ===")
if os.path.exists(hs_path):
    hs_folders = sorted([d for d in os.listdir(hs_path) if os.path.isdir(os.path.join(hs_path, d))])
    print(f"Total HS folders: {len(hs_folders)}")
    print(f"First few: {hs_folders[:5]}")
else:
    print("HS folder tidak ditemukan!")

=== CVA Folders ===
Total CVA folders: 49
First few: ['CVA_1', 'CVA_10', 'CVA_11', 'CVA_12', 'CVA_13']

=== HS Folders ===
Total HS folders: 73
First few: ['HS_1', 'HS_10', 'HS_11', 'HS_12', 'HS_13']


In [3]:
import os
import json

# Cari file meta di folder CVA dan HS
meta_files = []

# Cari di CVA
for root, dirs, files in os.walk(cva_path):
    for f in files:
        if f.endswith("_meta.json"):
            meta_files.append(os.path.join(root, f))

# Cari di HS
for root, dirs, files in os.walk(hs_path):
    for f in files:
        if f.endswith("_meta.json"):
            meta_files.append(os.path.join(root, f))

print(f"Total meta files ditemukan: {len(meta_files)}")
for p in meta_files[:5]:
    print(p)

Total meta files ditemukan: 488
c:\Users\fizel\OneDrive\Documents\Proyek Penelitian_Progress\CVA\CVA_1\CVA_1_1\CVA_1_1_meta.json
c:\Users\fizel\OneDrive\Documents\Proyek Penelitian_Progress\CVA\CVA_1\CVA_1_2\CVA_1_2_meta.json
c:\Users\fizel\OneDrive\Documents\Proyek Penelitian_Progress\CVA\CVA_10\CVA_10_1\CVA_10_1_meta.json
c:\Users\fizel\OneDrive\Documents\Proyek Penelitian_Progress\CVA\CVA_10\CVA_10_2\CVA_10_2_meta.json
c:\Users\fizel\OneDrive\Documents\Proyek Penelitian_Progress\CVA\CVA_11\CVA_11_1\CVA_11_1_meta.json


In [4]:
# Tampilkan struktur direktori lokal
print("=== STRUKTUR FOLDER CVA ===")
for root, dirs, files in os.walk(cva_path):
    depth = root.replace(cva_path, '').count(os.sep)
    if depth > 2:  # Batasi kedalaman
        continue
    indent = '  ' * depth
    print(f"{indent}📁 {os.path.basename(root)}/")
    for f in files[:3]:
        print(f"{indent}   📄 {f}")
    if len(files) > 3:
        print(f"{indent}   ... (+{len(files)-3} files)")

print("\n=== STRUKTUR FOLDER HS ===")
for root, dirs, files in os.walk(hs_path):
    depth = root.replace(hs_path, '').count(os.sep)
    if depth > 2:  # Batasi kedalaman
        continue
    indent = '  ' * depth
    print(f"{indent}📁 {os.path.basename(root)}/")
    for f in files[:3]:
        print(f"{indent}   📄 {f}")
    if len(files) > 3:
        print(f"{indent}   ... (+{len(files)-3} files)")

=== STRUKTUR FOLDER CVA ===
📁 CVA/
  📁 CVA_1/
    📁 CVA_1_1/
       📄 CVA_1_1_meta.json
       📄 CVA_1_1_plot.png
       📄 CVA_1_1_processed_data.txt
       ... (+4 files)
    📁 CVA_1_2/
       📄 CVA_1_2_meta.json
       📄 CVA_1_2_plot.png
       📄 CVA_1_2_processed_data.txt
       ... (+4 files)
  📁 CVA_10/
    📁 CVA_10_1/
       📄 CVA_10_1_meta.json
       📄 CVA_10_1_plot.png
       📄 CVA_10_1_processed_data.txt
       ... (+4 files)
    📁 CVA_10_2/
       📄 CVA_10_2_meta.json
       📄 CVA_10_2_plot.png
       📄 CVA_10_2_processed_data.txt
       ... (+4 files)
  📁 CVA_11/
    📁 CVA_11_1/
       📄 CVA_11_1_meta.json
       📄 CVA_11_1_plot.png
       📄 CVA_11_1_processed_data.txt
       ... (+4 files)
    📁 CVA_11_2/
       📄 CVA_11_2_meta.json
       📄 CVA_11_2_plot.png
       📄 CVA_11_2_processed_data.txt
       ... (+4 files)
  📁 CVA_12/
    📁 CVA_12_1/
       📄 CVA_12_1_meta.json
       📄 CVA_12_1_plot.png
       📄 CVA_12_1_processed_data.txt
       ... (+4 files)
    📁 CVA_12_2/


---
## TAHAP 2 — PIPELINE LANJUTAN
### *(Lanjutan dari LoadDataset — variabel `meta_files`, `base_path`, `cva_path`, `hs_path` sudah tersedia)*

Pipeline:
```
meta_files (SUDAH ADA)
→ Preprocessing → Label Generation → Sliding Window
→ BiTCN–BiGRU–CrossAttention Training
→ Event Detection (LHS, LTO, RHS, RTO)
→ Timestamp Extraction → 20 Gait Parameters → CSV
```

**Label mapping:**
- `leftGaitEvents[i]  = [LTO, LHS]`
- `rightGaitEvents[i] = [RTO, RHS]`
- Sequence label: `0=None, 1=LHS, 2=LTO, 3=RHS, 4=RTO`


### CELL 2.1 — Tambahan Import Library

In [5]:
# Tambahan library (yang belum di-import di TAHAP 1)
import warnings, logging, time
from pathlib import Path
from collections import defaultdict
from copy import deepcopy

import matplotlib.pyplot as plt
import seaborn as sns
from scipy.signal import butter, filtfilt, find_peaks
from scipy.ndimage import label as ndimage_label
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import (classification_report, confusion_matrix, f1_score)
from tqdm.auto import tqdm

import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import Dataset, DataLoader

# warnings.filterwarnings('ignore')

# print(f"PyTorch  : {torch.__version__}")
# print(f"CUDA     : {torch.cuda.is_available()}")
# device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
# print(f"Device   : {device}")

warnings.filterwarnings('ignore')

print(f"PyTorch  : {torch.__version__}")
print(f"CUDA     : {torch.cuda.is_available()}")

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f"Device   : {device}")

if device.type == 'cuda':
    print(f"GPU Name : {torch.cuda.get_device_name(0)}")
    print(f"VRAM     : {torch.cuda.get_device_properties(0).total_memory / 1e9:.2f} GB")
    
    # Biar training lebih cepat (cuDNN optimization)
    torch.backends.cudnn.benchmark = True
else:
    print("⚠️ GPU tidak terdeteksi, masih pakai CPU")


c:\Users\fizel\OneDrive\Documents\Proyek Penelitian_Progress\env\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


PyTorch  : 2.11.0+cpu
CUDA     : False
Device   : cpu
⚠️ GPU tidak terdeteksi, masih pakai CPU


### CELL 2.2 — Konfigurasi

In [6]:
# ═══════════════════════════════════════════════════════════════
#  KONFIGURASI — ubah di sini jika perlu
#  base_path sudah ada dari TAHAP 1 (LoadDataset)
# ═══════════════════════════════════════════════════════════════

# Output directories
OUTPUT_DIR = os.path.join(base_path, 'outputs')
PRED_DIR   = os.path.join(OUTPUT_DIR, 'event_predictions')
MODEL_PATH = os.path.join(OUTPUT_DIR, 'best_model.pt')
METRICS_PATH = os.path.join(OUTPUT_DIR, 'test_metrics.json')
CSV_PATH   = os.path.join(OUTPUT_DIR, 'gait_parameters_all_trials.csv')
SCALER_PATH  = os.path.join(OUTPUT_DIR, 'scaler.pkl')

for d in [OUTPUT_DIR, PRED_DIR]:
    os.makedirs(d, exist_ok=True)

# ── Label / Event ────────────────────────────────────────────
EVENT_TOLERANCE = 5          # ±N sample di sekitar event
NUM_CLASSES     = 5          # 0=None,1=LHS,2=LTO,3=RHS,4=RTO
CLASS_NAMES     = ['None', 'LHS', 'LTO', 'RHS', 'RTO']
EVENT_MAP       = {1:'LHS', 2:'LTO', 3:'RHS', 4:'RTO'}

# ── Windowing ────────────────────────────────────────────────
WINDOW_SIZE = 256
STRIDE      = 64

# ── Model hyperparameters ────────────────────────────────────
TCN_CHANNELS = [64, 128, 128]   # output channels per TCN layer
TCN_KERNEL   = 3
TCN_DROPOUT  = 0.2
GRU_HIDDEN   = 128
GRU_LAYERS   = 2
ATTN_HEADS   = 4
ATTN_DROPOUT = 0.1

# ── Training ─────────────────────────────────────────────────
BATCH_SIZE   = 32
EPOCHS       = 80
LR           = 1e-3
WEIGHT_DECAY = 1e-4
PATIENCE     = 12            # early stopping

# ── IMU axis mapping untuk Lower Back ────────────────────────
# Sesuaikan jika nama kolom berbeda di dataset kamu
# Fallback otomatis ke 3 kolom pertama jika tidak ditemukan
AXIS_CC = 'Acc_X'   # craniocaudal
AXIS_ML = 'Acc_Y'   # mediolateral
AXIS_AP = 'Acc_Z'   # anteroposterior

# ── Seed ─────────────────────────────────────────────────────
SEED = 42
np.random.seed(SEED)
torch.manual_seed(SEED)

print("Konfigurasi selesai.")
print(f"  OUTPUT_DIR  : {OUTPUT_DIR}")
print(f"  WINDOW_SIZE : {WINDOW_SIZE} | STRIDE : {STRIDE}")
print(f"  EPOCHS      : {EPOCHS} | PATIENCE : {PATIENCE}")


Konfigurasi selesai.
  OUTPUT_DIR  : c:\Users\fizel\OneDrive\Documents\Proyek Penelitian_Progress\outputs
  WINDOW_SIZE : 256 | STRIDE : 64
  EPOCHS      : 80 | PATIENCE : 12


### CELL 2.3 — Bangun `ALL_TRIALS` dari `meta_files`

In [7]:
# ───────────────────────────────────────────────────────────────
#  Gunakan variabel `meta_files` yang SUDAH ADA dari TAHAP 1.
#  Bangun list of dict ALL_TRIALS tanpa menyentuh dataset.
# ───────────────────────────────────────────────────────────────

def build_trial_list(meta_files_list: list) -> list:
    """
    Dari list path meta.json, bangun ALL_TRIALS.
    Setiap elemen berisi:
      - meta_path   : path ke *_meta.json
      - data_path   : path ke *_processed_data.txt
      - group       : 'CVA' atau 'HS'
      - subject_id  : e.g. 'CVA_1'
      - trial_id    : e.g. 'CVA_1_1'
    """
    trials = []
    for meta_path in meta_files_list:
        meta_p = Path(meta_path)
        # Derive data path: ganti _meta.json -> _processed_data.txt
        data_path = str(meta_p).replace('_meta.json', '_processed_data.txt')
        if not os.path.exists(data_path):
            # Coba cari di folder yang sama
            candidates = list(meta_p.parent.glob('*_processed_data.txt'))
            if not candidates:
                continue
            data_path = str(candidates[0])

        # Tentukan group dari path
        parts = meta_p.parts
        group = 'CVA' if any('CVA' in p and not '_' in p.replace('CVA','') for p in parts) else 'HS'
        # Lebih robust: cek apakah 'CVA' atau 'HS' ada sebagai folder di path
        group = 'unknown'
        for p in parts:
            if p == 'CVA':
                group = 'CVA'; break
            if p == 'HS':
                group = 'HS'; break

        trial_id   = meta_p.parent.name    # e.g. CVA_1_1
        subject_id = meta_p.parent.parent.name  # e.g. CVA_1

        trials.append({
            'meta_path' : str(meta_path),
            'data_path' : str(data_path),
            'group'     : group,
            'subject_id': subject_id,
            'trial_id'  : trial_id,
        })
    return trials

ALL_TRIALS = build_trial_list(meta_files)

# Verifikasi
from collections import Counter
gc = Counter(t['group'] for t in ALL_TRIALS)
print(f"ALL_TRIALS dibangun dari meta_files:")
print(f"  Total : {len(ALL_TRIALS)}")
print(f"  CVA   : {gc.get('CVA', 0)}")
print(f"  HS    : {gc.get('HS', 0)}")
print(f"\nContoh 3 trials:")
for t in ALL_TRIALS[:3]:
    print(f"  {t['trial_id']:15s} | group={t['group']} | data_exists={os.path.exists(t['data_path'])}")


ALL_TRIALS dibangun dari meta_files:
  Total : 488
  CVA   : 128
  HS    : 360

Contoh 3 trials:
  CVA_1_1         | group=CVA | data_exists=True
  CVA_1_2         | group=CVA | data_exists=True
  CVA_10_1        | group=CVA | data_exists=True


### CELL 2.4 — Fungsi Utilitas (Loader, Filter, Column Selector)

In [8]:
# ═══════════════════════════════════════════════════════════════
#  UTILITY FUNCTIONS
#  Semua modular, tidak hardcode nama kolom
# ═══════════════════════════════════════════════════════════════

# ── 1. Flexible TXT loader ────────────────────────────────────
def load_txt_flexible(filepath: str) -> pd.DataFrame:
    """
    Baca processed_data.txt dengan auto-detect delimiter.
    Coba tab, comma, semicolon, whitespace.
    """
    for sep in ['	', ',', ';', r'\s+']:
        try:
            df = pd.read_csv(filepath, sep=sep, engine='python',
                             on_bad_lines='skip')
            if df.shape[1] >= 3:
                return df
        except Exception:
            continue
    raise ValueError(f"Gagal parsing: {filepath}")

# ── 2. Clean column names ─────────────────────────────────────
def clean_columns(df: pd.DataFrame) -> pd.DataFrame:
    df.columns = [str(c).strip().replace(' ', '_') for c in df.columns]
    return df

# ── 3. Select IMU channels ────────────────────────────────────
SENSOR_KEYWORDS = ['acc', 'gyr', 'gyro', 'accel', 'mag']

def select_imu_channels(df: pd.DataFrame) -> tuple:
    """
    Pilih kolom numerik yang relevan dengan IMU.
    Prioritas: kolom mengandung kata kunci sensor (acc/gyr).
    Fallback: semua kolom numerik.
    Return: (array (T,C), list kolom names)
    """
    numeric_cols = df.select_dtypes(include=[np.number]).columns.tolist()
    sensor_cols  = [c for c in numeric_cols
                    if any(kw in c.lower() for kw in SENSOR_KEYWORDS)]
    chosen = sensor_cols if len(sensor_cols) >= 3 else numeric_cols
    if len(chosen) < 1:
        raise ValueError("Tidak cukup kolom numerik.")
    return df[chosen].fillna(0).values.astype(np.float32), chosen

# ── 4. Replace nan/inf ────────────────────────────────────────
def safe_arr(arr: np.ndarray) -> np.ndarray:
    return np.where(np.isfinite(arr), arr, 0.0).astype(np.float32)

# ── 5. Get Lower Back axis columns ────────────────────────────
def get_lb_axis_map(df: pd.DataFrame) -> dict:
    """
    Cari kolom Lower Back untuk ACC dan GYR.
    Return dict {'CC':col, 'ML':col, 'AP':col, 'GyrX':col, ...}
    Fallback ke 3 kolom acc pertama jika tidak ada label LB eksplisit.
    """
    cols   = df.columns.tolist()
    num_c  = df.select_dtypes(np.number).columns.tolist()
    result = {}

    # Coba match dari config (AXIS_CC, AXIS_ML, AXIS_AP)
    if AXIS_CC in df.columns: result['CC'] = AXIS_CC
    if AXIS_ML in df.columns: result['ML'] = AXIS_ML
    if AXIS_AP in df.columns: result['AP'] = AXIS_AP

    # Cari kolom LB / lower_back secara eksplisit
    lb_acc = [c for c in num_c if 'lb' in c.lower() and 'acc' in c.lower()]
    lb_gyr = [c for c in num_c if 'lb' in c.lower() and ('gyr' in c.lower() or 'gyro' in c.lower())]
    all_acc= [c for c in num_c if 'acc' in c.lower()]
    all_gyr= [c for c in num_c if 'gyr' in c.lower() or 'gyro' in c.lower()]

    acc_pool = lb_acc if lb_acc else all_acc
    gyr_pool = lb_gyr if lb_gyr else all_gyr

    if 'CC' not in result and len(acc_pool) >= 1: result['CC'] = acc_pool[0]
    if 'ML' not in result and len(acc_pool) >= 2: result['ML'] = acc_pool[1]
    if 'AP' not in result and len(acc_pool) >= 3: result['AP'] = acc_pool[2]
    if len(gyr_pool) >= 1: result['GyrX'] = gyr_pool[0]
    if len(gyr_pool) >= 2: result['GyrY'] = gyr_pool[1]
    if len(gyr_pool) >= 3: result['GyrZ'] = gyr_pool[2]

    return result

# ── 6. Butterworth low-pass filter ───────────────────────────
def butter_lowpass(data: np.ndarray, cutoff=20.0, fs=100.0, order=4):
    nyq = fs / 2.0
    b, a = butter(order, min(cutoff/nyq, 0.99), btype='low', analog=False)
    if data.ndim == 1:
        return filtfilt(b, a, data)
    return np.apply_along_axis(lambda c: filtfilt(b, a, c), 0, data)

print("Utility functions siap.")


Utility functions siap.


### CELL 2.5 — Load Satu Trial (Demo)

In [9]:
def load_trial(trial_info: dict) -> dict:
    """
    Load satu trial: baca meta.json + processed_data.txt.
    Input : dict dari ALL_TRIALS
    Output: dict {meta, signal (T,C), columns, df, trial_id, group}
    Tidak mengubah dataset apapun.
    """
    # Load meta
    with open(trial_info['meta_path'], 'r') as f:
        meta = json.load(f)

    # Load signal
    df = load_txt_flexible(trial_info['data_path'])
    df = clean_columns(df)
    df = df.dropna(how='all').reset_index(drop=True)

    X, chosen_cols = select_imu_channels(df)
    X = safe_arr(X)

    return {
        'meta'    : meta,
        'signal'  : X,
        'columns' : chosen_cols,
        'df'      : df,
        'trial_id': trial_info['trial_id'],
        'group'   : trial_info['group'],
    }

# Demo load satu trial
if ALL_TRIALS:
    _demo = load_trial(ALL_TRIALS[0])
    print(f"Trial     : {_demo['trial_id']} | Group: {_demo['group']}")
    print(f"Signal    : shape={_demo['signal'].shape}  dtype={_demo['signal'].dtype}")
    print(f"Channels  : {_demo['columns']}")
    print(f"Freq      : {_demo['meta'].get('freq','?')} Hz")
    print(f"leftGaitEvents  (sample) : {_demo['meta'].get('leftGaitEvents',[])[:2]}")
    print(f"rightGaitEvents (sample) : {_demo['meta'].get('rightGaitEvents',[])[:2]}")


Trial     : CVA_1_1 | Group: CVA
Signal    : shape=(3639, 36)  dtype=float32
Channels  : ['HE_Acc_X', 'HE_Acc_Y', 'HE_Acc_Z', 'HE_FreeAcc_X', 'HE_FreeAcc_Y', 'HE_FreeAcc_Z', 'HE_Gyr_X', 'HE_Gyr_Y', 'HE_Gyr_Z', 'LB_Acc_X', 'LB_Acc_Y', 'LB_Acc_Z', 'LB_FreeAcc_X', 'LB_FreeAcc_Y', 'LB_FreeAcc_Z', 'LB_Gyr_X', 'LB_Gyr_Y', 'LB_Gyr_Z', 'LF_Acc_X', 'LF_Acc_Y', 'LF_Acc_Z', 'LF_FreeAcc_X', 'LF_FreeAcc_Y', 'LF_FreeAcc_Z', 'LF_Gyr_X', 'LF_Gyr_Y', 'LF_Gyr_Z', 'RF_Acc_X', 'RF_Acc_Y', 'RF_Acc_Z', 'RF_FreeAcc_X', 'RF_FreeAcc_Y', 'RF_FreeAcc_Z', 'RF_Gyr_X', 'RF_Gyr_Y', 'RF_Gyr_Z']
Freq      : 100.0 Hz
leftGaitEvents  (sample) : [[487, 509], [620, 666]]
rightGaitEvents (sample) : [[407, 464], [540, 606]]


### CELL 2.6 — Label Generation dari meta.json

In [8]:
def generate_labels(T: int, meta: dict,
                     tolerance: int = EVENT_TOLERANCE) -> np.ndarray:
    """
    Buat array label panjang T.
    Label: 0=None, 1=LHS, 2=LTO, 3=RHS, 4=RTO

    Format meta.json:
      leftGaitEvents[i]  = [LTO_idx, LHS_idx]
      rightGaitEvents[i] = [RTO_idx, RHS_idx]

    Setiap event diberi toleransi window ±tolerance sample.
    """
    labels = np.zeros(T, dtype=np.int64)

    def mark(label_val, idx):
        lo = max(0, idx - tolerance)
        hi = min(T, idx + tolerance + 1)
        labels[lo:hi] = label_val

    for pair in meta.get('leftGaitEvents', []):
        if len(pair) >= 2:
            lto_i, lhs_i = int(pair[0]), int(pair[1])
            if 0 <= lto_i < T: mark(2, lto_i)   # LTO = class 2
            if 0 <= lhs_i < T: mark(1, lhs_i)   # LHS = class 1

    for pair in meta.get('rightGaitEvents', []):
        if len(pair) >= 2:
            rto_i, rhs_i = int(pair[0]), int(pair[1])
            if 0 <= rto_i < T: mark(4, rto_i)   # RTO = class 4
            if 0 <= rhs_i < T: mark(3, rhs_i)   # RHS = class 3

    return labels

# Demo
if ALL_TRIALS:
    _T  = _demo['signal'].shape[0]
    _y  = generate_labels(_T, _demo['meta'])
    uni, cnt = np.unique(_y, return_counts=True)
    print(f"Label distribution — Trial {_demo['trial_id']} (T={_T}):")
    for u, c in zip(uni, cnt):
        print(f"  {CLASS_NAMES[u]:6s} (label={u}) : {c:6d} frames  ({100*c/_T:.2f}%)")


NameError: name '_demo' is not defined

### CELL 2.7 — Preprocessing Signal

In [11]:
def preprocess_signal(X: np.ndarray, freq: float = 100.0,
                       cutoff: float = 20.0) -> np.ndarray:
    """
    Preprocessing sinyal IMU:
      1. Replace NaN/Inf dengan 0
      2. Butterworth low-pass filter (cutoff=20 Hz)
    Normalisasi (z-score) dilakukan secara global setelah split.
    """
    X = safe_arr(X).copy()
    if X.shape[0] > 20:
        try:
            X = butter_lowpass(X, cutoff=cutoff, fs=freq)
        except Exception:
            pass   # lewati filter jika data terlalu pendek
    return X.astype(np.float32)

# StandardScaler — di-fit HANYA dari training set (diisi di CELL 2.9)
GLOBAL_SCALER = None

def fit_scaler(X_list: list) -> StandardScaler:
    """Fit StandardScaler dari list array (T,C)."""
    all_data = np.vstack(X_list)
    sc = StandardScaler()
    sc.fit(all_data)
    return sc

def apply_scaler(X: np.ndarray, scaler) -> np.ndarray:
    return scaler.transform(X).astype(np.float32)

print("Preprocessing functions siap.")


Preprocessing functions siap.


### CELL 2.8 — Sliding Window

In [12]:
def make_windows(X: np.ndarray, y: np.ndarray,
                  window_size: int = WINDOW_SIZE,
                  stride: int = STRIDE) -> tuple:
    """
    Buat sliding window dari satu trial.
    X: (T, C)  →  X_wins: (N, window_size, C)
    y: (T,)    →  y_wins: (N, window_size)
    Trial lebih pendek dari window_size akan di-pad dengan nol.
    """
    T = X.shape[0]
    if T < window_size:
        pad = window_size - T
        X   = np.vstack([X, np.zeros((pad, X.shape[1]), dtype=np.float32)])
        y   = np.concatenate([y, np.zeros(pad, dtype=np.int64)])
        T   = window_size

    X_wins, y_wins = [], []
    for start in range(0, T - window_size + 1, stride):
        end = start + window_size
        X_wins.append(X[start:end])
        y_wins.append(y[start:end])

    return (np.array(X_wins, dtype=np.float32),
            np.array(y_wins, dtype=np.int64))

print(f"Sliding window siap: WINDOW_SIZE={WINDOW_SIZE}, STRIDE={STRIDE}")


Sliding window siap: WINDOW_SIZE=256, STRIDE=64


### CELL 2.9 — Subject-Level Train/Val/Test Split & Load Semua Trial

In [13]:
def subject_split(trials: list, train_r=0.70, val_r=0.15,
                   seed=SEED) -> tuple:
    """
    Split berdasarkan SUBJECT (bukan window) → cegah data leakage.
    CVA dan HS di-split terpisah agar distribusi terjaga.
    """
    np.random.seed(seed)
    tr, va, te = [], [], []

    for group in ['CVA', 'HS']:
        g_trials = [t for t in trials if t['group'] == group]
        subjects = sorted({t['subject_id'] for t in g_trials})
        np.random.shuffle(subjects)
        n      = len(subjects)
        n_tr   = int(n * train_r)
        n_va   = int(n * val_r)
        s_tr   = set(subjects[:n_tr])
        s_va   = set(subjects[n_tr:n_tr+n_va])
        s_te   = set(subjects[n_tr+n_va:])
        for t in g_trials:
            s = t['subject_id']
            if   s in s_tr: tr.append(t)
            elif s in s_va: va.append(t)
            else:           te.append(t)

    print(f"Split (subject-level): train={len(tr)}  val={len(va)}  test={len(te)}")
    return tr, va, te

TRAIN_TRIALS, VAL_TRIALS, TEST_TRIALS = subject_split(ALL_TRIALS)

# ── Load + preprocess semua trials ───────────────────────────
def load_all(trial_list: list, desc='Loading') -> list:
    data, errs = [], []
    for info in tqdm(trial_list, desc=desc):
        try:
            t    = load_trial(info)
            freq = float(t['meta'].get('freq', 100.0))
            X    = preprocess_signal(t['signal'], freq=freq)
            y    = generate_labels(X.shape[0], t['meta'])
            data.append({
                'X': X, 'y': y,
                'meta'    : t['meta'],
                'df'      : t['df'],
                'trial_id': info['trial_id'],
                'group'   : info['group'],
                'subject' : info['subject_id'],
            })
        except Exception as e:
            errs.append((info['trial_id'], str(e)))
    if errs:
        print(f"[WARN] {len(errs)} trial gagal load:")
        for tid, err in errs[:5]: print(f"  {tid}: {err}")
    return data

print("\nLoading training set...")
TRAIN_DATA = load_all(TRAIN_TRIALS, desc='Train')
print(f"  → {len(TRAIN_DATA)} trials")

# Fit global scaler HANYA dari training data
GLOBAL_SCALER = fit_scaler([d['X'] for d in TRAIN_DATA])
for d in TRAIN_DATA:
    d['X'] = apply_scaler(d['X'], GLOBAL_SCALER)

print("Loading val & test set...")
VAL_DATA  = load_all(VAL_TRIALS,  desc='Val')
TEST_DATA = load_all(TEST_TRIALS, desc='Test')
for d in VAL_DATA + TEST_DATA:
    d['X'] = apply_scaler(d['X'], GLOBAL_SCALER)

# Simpan scaler
with open(SCALER_PATH, 'wb') as f:
    pickle.dump(GLOBAL_SCALER, f)

N_CHANNELS = TRAIN_DATA[0]['X'].shape[1] if TRAIN_DATA else 6
print(f"\nInput channels  : {N_CHANNELS}")
print(f"Scaler saved    : {SCALER_PATH}")


Split (subject-level): train=362  val=56  test=70

Loading training set...


Train: 100%|██████████| 362/362 [00:43<00:00,  8.39it/s]


  → 362 trials
Loading val & test set...


Test: 100%|██████████| 70/70 [00:07<00:00,  9.56it/s]



Input channels  : 36
Scaler saved    : c:\Users\fizel\OneDrive\Documents\Proyek Penelitian_Progress\outputs\scaler.pkl


### CELL 2.10 — PyTorch Dataset & DataLoader

In [14]:
class GaitWindowDataset(Dataset):
    """
    Sliding window dataset dari list data trial.
    Setiap item: (X_window: FloatTensor (W,C), y_window: LongTensor (W,))
    """
    def __init__(self, data_list: list,
                 window_size=WINDOW_SIZE, stride=STRIDE):
        self.samples = []
        for d in data_list:
            X_w, y_w = make_windows(d['X'], d['y'], window_size, stride)
            for i in range(len(X_w)):
                self.samples.append((X_w[i], y_w[i]))

    def __len__(self):
        return len(self.samples)

    def __getitem__(self, idx):
        X, y = self.samples[idx]
        return (torch.from_numpy(X).float(),
                torch.from_numpy(y).long())

train_ds = GaitWindowDataset(TRAIN_DATA)
val_ds   = GaitWindowDataset(VAL_DATA)
test_ds  = GaitWindowDataset(TEST_DATA)

print(f"Windows: train={len(train_ds)}  val={len(val_ds)}  test={len(test_ds)}")

# ── Class weights (inverse frequency) ────────────────────────
label_counts = np.zeros(NUM_CLASSES, dtype=np.float64)
for X_, y_ in train_ds:
    vals, cnts = torch.unique(y_, return_counts=True)
    for v, c in zip(vals.numpy(), cnts.numpy()):
        label_counts[int(v)] += int(c)

label_counts  = np.where(label_counts == 0, 1, label_counts)
class_weights = 1.0 / label_counts
class_weights = class_weights / class_weights.sum() * NUM_CLASSES
CLASS_WEIGHTS = torch.tensor(class_weights, dtype=torch.float32).to(device)

print("\nClass weights (inverse frequency):")
for i, w in enumerate(CLASS_WEIGHTS.cpu().numpy()):
    print(f"  {CLASS_NAMES[i]:6s}: count={int(label_counts[i]):9d}  weight={w:.4f}")

train_loader = DataLoader(train_ds, batch_size=BATCH_SIZE, shuffle=True,  num_workers=0)
val_loader   = DataLoader(val_ds,   batch_size=BATCH_SIZE, shuffle=False, num_workers=0)
test_loader  = DataLoader(test_ds,  batch_size=BATCH_SIZE, shuffle=False, num_workers=0)
print(f"\nDataLoaders siap — batch_size={BATCH_SIZE}")


Windows: train=13853  val=1908  test=2675

Class weights (inverse frequency):
  None  : count=  2434806  weight=0.1387
  LHS   : count=   273677  weight=1.2339
  LTO   : count=   277716  weight=1.2160
  RHS   : count=   278446  weight=1.2128
  RTO   : count=   281723  weight=1.1987

DataLoaders siap — batch_size=32


### CELL 2.11 — BiTCN–BiGRU–Cross Attention Model

Arsitektur berdasarkan Xu et al. (2025) *Bioengineering* 12(5):491:

```
Input (B, T, C)
→ BiTCN (forward + backward dilated TCN, concat)
→ BatchNorm + ReLU
→ BiGRU (2 layer)
→ Cross Attention (MultiheadAttention, Q=K=V)
→ LayerNorm + residual
→ Linear → (B, T, 5)
```


In [15]:
# ── 1. Causal Dilated Convolution ────────────────────────────
class CausalConv1d(nn.Module):
    """Dilated causal 1D conv: tidak bocorkan info future."""
    def __init__(self, in_ch, out_ch, kernel_size, dilation):
        super().__init__()
        self.pad  = (kernel_size - 1) * dilation
        self.conv = nn.Conv1d(in_ch, out_ch, kernel_size,
                              padding=self.pad, dilation=dilation)

    def forward(self, x):               # x: (B, C, T)
        return self.conv(x)[:, :, :x.size(2)]   # trim padding

# ── 2. TCN Residual Block ─────────────────────────────────────
class TCNBlock(nn.Module):
    """
    Residual block:
      CausalConv → BN → ReLU → Dropout
      CausalConv → BN → ReLU → Dropout
      + residual (1x1 conv jika channel berubah)
    """
    def __init__(self, in_ch, out_ch, kernel_size, dilation,
                 dropout=0.2):
        super().__init__()
        self.conv1 = CausalConv1d(in_ch,  out_ch, kernel_size, dilation)
        self.conv2 = CausalConv1d(out_ch, out_ch, kernel_size, dilation)
        self.bn1   = nn.BatchNorm1d(out_ch)
        self.bn2   = nn.BatchNorm1d(out_ch)
        self.relu  = nn.ReLU()
        self.drop  = nn.Dropout(dropout)
        self.proj  = (nn.Conv1d(in_ch, out_ch, 1)
                      if in_ch != out_ch else nn.Identity())

    def forward(self, x):               # x: (B, C, T)
        res = self.proj(x)
        out = self.drop(self.relu(self.bn1(self.conv1(x))))
        out = self.drop(self.relu(self.bn2(self.conv2(out))))
        return self.relu(out + res)

# ── 3. Bidirectional TCN ──────────────────────────────────────
class BiTCN(nn.Module):
    """
    Forward TCN (dilation 1,2,4,...) + Backward TCN (flip), concat.
    Output dim = channels[-1] * 2
    """
    def __init__(self, in_ch, channels: list,
                 kernel_size=3, dropout=0.2):
        super().__init__()
        self.fwd = nn.ModuleList()
        self.bwd = nn.ModuleList()
        ch = in_ch
        for i, out_ch in enumerate(channels):
            dil = 2 ** i
            self.fwd.append(TCNBlock(ch, out_ch, kernel_size, dil, dropout))
            self.bwd.append(TCNBlock(ch, out_ch, kernel_size, dil, dropout))
            ch = out_ch
        self.out_channels = channels[-1] * 2

    def forward(self, x):               # x: (B, T, C_in)
        x_t  = x.transpose(1, 2)        # → (B, C_in, T)
        f    = x_t
        b    = torch.flip(x_t, dims=[2])
        for fb, bb in zip(self.fwd, self.bwd):
            f = fb(f)
            b = bb(b)
        b    = torch.flip(b, dims=[2])
        out  = torch.cat([f, b], dim=1) # (B, ch*2, T)
        return out.transpose(1, 2)      # (B, T, ch*2)

# ── 4. Full Model ─────────────────────────────────────────────
class BiTCN_BiGRU_CrossAttn(nn.Module):
    """
    BiTCN → BN+ReLU → BiGRU → CrossAttention → Linear
    Output: (B, T, num_classes)
    """
    def __init__(self, in_channels: int,
                 tcn_channels=TCN_CHANNELS,
                 tcn_kernel=TCN_KERNEL,
                 tcn_dropout=TCN_DROPOUT,
                 gru_hidden=GRU_HIDDEN,
                 gru_layers=GRU_LAYERS,
                 attn_heads=ATTN_HEADS,
                 attn_dropout=ATTN_DROPOUT,
                 num_classes=NUM_CLASSES):
        super().__init__()

        # BiTCN
        self.bitcn    = BiTCN(in_channels, tcn_channels,
                               tcn_kernel, tcn_dropout)
        tcn_dim       = self.bitcn.out_channels   # ch*2

        # BN + ReLU setelah BiTCN
        self.bn_tcn   = nn.BatchNorm1d(tcn_dim)
        self.relu     = nn.ReLU()

        # BiGRU
        self.bigru    = nn.GRU(
            input_size   = tcn_dim,
            hidden_size  = gru_hidden,
            num_layers   = gru_layers,
            batch_first  = True,
            bidirectional= True,
            dropout      = tcn_dropout if gru_layers > 1 else 0.0
        )
        gru_dim = gru_hidden * 2           # bidirectional output

        # Cross Attention (self-attention: Q=K=V = BiGRU output)
        self.cross_attn = nn.MultiheadAttention(
            embed_dim   = gru_dim,
            num_heads   = attn_heads,
            dropout     = attn_dropout,
            batch_first = True
        )
        self.attn_norm  = nn.LayerNorm(gru_dim)

        # Classifier head
        self.classifier = nn.Linear(gru_dim, num_classes)

    def forward(self, x):               # x: (B, T, C)
        # BiTCN
        tcn_out = self.bitcn(x)          # (B, T, tcn_dim)
        # BN: transpose → BN → transpose
        tcn_out = self.bn_tcn(
            tcn_out.transpose(1,2)
        ).transpose(1,2)
        tcn_out = self.relu(tcn_out)
        # BiGRU
        gru_out, _ = self.bigru(tcn_out) # (B, T, gru_dim)
        # Cross Attention (residual connection)
        attn_out, _ = self.cross_attn(gru_out, gru_out, gru_out)
        attn_out     = self.attn_norm(attn_out + gru_out)
        # Classifier
        return self.classifier(attn_out) # (B, T, num_classes)

# ── Instantiate ───────────────────────────────────────────────
model = BiTCN_BiGRU_CrossAttn(in_channels=N_CHANNELS).to(device)

n_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print(f"Model     : BiTCN_BiGRU_CrossAttn")
print(f"Params    : {n_params:,}")
print(f"In channels: {N_CHANNELS}")

# Verifikasi forward pass
_xb = torch.randn(2, WINDOW_SIZE, N_CHANNELS).to(device)
_out = model(_xb)
print(f"Forward pass OK: input={_xb.shape} → output={_out.shape}")
del _xb, _out


Model     : BiTCN_BiGRU_CrossAttn
Params    : 1,266,053
In channels: 36
Forward pass OK: input=torch.Size([2, 256, 36]) → output=torch.Size([2, 256, 5])


### CELL 2.12 — Training Loop

In [20]:
def run_epoch(model, loader, optimizer, criterion,
              device, train=True):
    """Satu epoch training atau evaluasi."""
    model.train() if train else model.eval()
    total_loss = 0.0
    all_pred, all_true = [], []

    ctx = torch.enable_grad() if train else torch.no_grad()
    with ctx:
        for X_b, y_b in loader:
            X_b, y_b = X_b.to(device), y_b.to(device)
            logits   = model(X_b)                     # (B, T, C)
            loss     = criterion(
                logits.reshape(-1, NUM_CLASSES),
                y_b.reshape(-1))

            if train:
                optimizer.zero_grad()
                loss.backward()
                nn.utils.clip_grad_norm_(model.parameters(), 1.0)
                optimizer.step()

            total_loss += loss.item() * X_b.size(0)
            preds = logits.argmax(-1).reshape(-1).cpu().numpy()
            all_pred.append(preds)
            all_true.append(y_b.reshape(-1).cpu().numpy())

    avg_loss = total_loss / len(loader.dataset)
    all_pred = np.concatenate(all_pred)
    all_true = np.concatenate(all_true)
    macro_f1 = f1_score(all_true, all_pred, average='macro',
                         zero_division=0,
                         labels=list(range(NUM_CLASSES)))
    return avg_loss, macro_f1, all_true, all_pred

# ── Setup ─────────────────────────────────────────────────────
criterion = nn.CrossEntropyLoss(weight=CLASS_WEIGHTS)
optimizer = torch.optim.AdamW(model.parameters(), lr=LR,
                               weight_decay=WEIGHT_DECAY)


history       = defaultdict(list)
best_val_f1   = -1.0
patience_cnt  = 0
best_state    = None

print(f"Training — max {EPOCHS} epochs | early stop patience={PATIENCE}")
print("-" * 72)

for epoch in range(1, EPOCHS + 1):
    t0 = time.time()
    tr_loss, tr_f1, _, _   = run_epoch(model, train_loader,
                                        optimizer, criterion, device, train=True)
    va_loss, va_f1, _, _   = run_epoch(model, val_loader,
                                        optimizer, criterion, device, train=False)
    #scheduler.step(va_f1)
    dt = time.time() - t0

    history['train_loss'].append(tr_loss)
    history['val_loss'].append(va_loss)
    history['train_f1'].append(tr_f1)
    history['val_f1'].append(va_f1)

    tag = ''
    if va_f1 > best_val_f1:
        best_val_f1  = va_f1
        best_state   = deepcopy(model.state_dict())
        patience_cnt = 0
        torch.save(best_state, MODEL_PATH)
        tag = '  ✓ best'
    else:
        patience_cnt += 1

    print(f"Epoch {epoch:3d}/{EPOCHS} | "
          f"TrLoss={tr_loss:.4f} TrF1={tr_f1:.4f} | "
          f"VaLoss={va_loss:.4f} VaF1={va_f1:.4f} | "
          f"{dt:.1f}s{tag}")

    if patience_cnt >= PATIENCE:
        print(f"Early stopping at epoch {epoch}.")
        break

model.load_state_dict(torch.load(MODEL_PATH, map_location=device))
print(f"\nTraining selesai. Best val macro F1 = {best_val_f1:.4f}")
print(f"Model disimpan : {MODEL_PATH}")


Training — max 80 epochs | early stop patience=12
------------------------------------------------------------------------
Epoch   1/80 | TrLoss=0.1169 TrF1=0.8808 | VaLoss=0.1305 VaF1=0.8976 | 684.9s  ✓ best
Epoch   2/80 | TrLoss=0.0934 TrF1=0.8982 | VaLoss=0.1265 VaF1=0.9045 | 646.8s  ✓ best
Epoch   3/80 | TrLoss=0.0815 TrF1=0.9086 | VaLoss=0.1241 VaF1=0.9010 | 651.4s
Epoch   4/80 | TrLoss=0.0722 TrF1=0.9168 | VaLoss=0.1305 VaF1=0.9083 | 655.7s  ✓ best
Epoch   5/80 | TrLoss=0.0659 TrF1=0.9228 | VaLoss=0.1350 VaF1=0.9170 | 752.0s  ✓ best
Epoch   6/80 | TrLoss=0.0608 TrF1=0.9283 | VaLoss=0.1363 VaF1=0.9155 | 725.1s
Epoch   7/80 | TrLoss=0.0571 TrF1=0.9327 | VaLoss=0.1388 VaF1=0.9176 | 679.0s  ✓ best
Epoch   8/80 | TrLoss=0.0538 TrF1=0.9359 | VaLoss=0.1473 VaF1=0.9206 | 748.1s  ✓ best
Epoch   9/80 | TrLoss=0.0505 TrF1=0.9398 | VaLoss=0.1572 VaF1=0.9187 | 808.3s
Epoch  10/80 | TrLoss=0.0463 TrF1=0.9442 | VaLoss=0.1680 VaF1=0.9184 | 698.6s
Epoch  11/80 | TrLoss=0.0462 TrF1=0.9445 | VaLoss

KeyboardInterrupt: 

### CELL 2.13 — Evaluasi Frame-Level (Test Set)

In [22]:
# Evaluasi pada test set
te_loss, te_f1, te_true, te_pred = run_epoch(
    model, test_loader, optimizer, criterion, device, train=False)

print(f"Test Loss    : {te_loss:.4f}")
print(f"Test Macro F1: {te_f1:.4f}")
print()
print("=== Per-Class Report ===")
print(classification_report(te_true, te_pred,
                             labels=list(range(NUM_CLASSES)),
                             target_names=CLASS_NAMES,
                             zero_division=0))

# Confusion matrix
cm = confusion_matrix(te_true, te_pred, labels=list(range(NUM_CLASSES)))
cm_df = pd.DataFrame(cm, index=CLASS_NAMES, columns=CLASS_NAMES)
print("Confusion Matrix (rows=True, cols=Predicted):")
print(cm_df)

# Per-class F1 per event
p, r, f, s = precision_recall_fscore_support(
    te_true, te_pred,
    labels=list(range(NUM_CLASSES)),
    zero_division=0)
from sklearn.metrics import precision_recall_fscore_support

# Simpan ke dict (akan di-update dengan event MAE nanti)
METRICS = {
    'test_loss'     : float(te_loss),
    'test_macro_f1' : float(te_f1),
    'per_class_f1'  : {CLASS_NAMES[i]: float(f[i]) for i in range(NUM_CLASSES)},
}


Test Loss    : 0.3776
Test Macro F1: 0.8977

=== Per-Class Report ===
              precision    recall  f1-score   support

        None       0.97      0.93      0.95    485954
         LHS       0.81      0.90      0.85     49017
         LTO       0.87      0.97      0.92     49832
         RHS       0.83      0.89      0.86     49672
         RTO       0.88      0.94      0.91     50325

    accuracy                           0.93    684800
   macro avg       0.87      0.93      0.90    684800
weighted avg       0.93      0.93      0.93    684800

Confusion Matrix (rows=True, cols=Predicted):
        None    LHS    LTO    RHS    RTO
None  453714  10197   6730   9137   6176
LHS     4862  43930     38      1    186
LTO     1202      0  48467    163      0
RHS     5376      1    237  44044     14
RTO     2626    203      0      7  47489


NameError: name 'precision_recall_fscore_support' is not defined

### CELL 2.14 — Event Timestamp Extraction

In [ ]:
def extract_events_from_sequence(pred_labels: np.ndarray,
                                   freq: float = 100.0,
                                   min_distance: int = 20) -> dict:
    """
    Ubah label sequence → timestamp event per kelas.

    Langkah:
      1. Untuk tiap kelas (1–4), cari segmen berturut-turut (connected components)
      2. Ambil midpoint setiap segmen sebagai event index
      3. Hapus event yang jaraknya < min_distance (ambil yang pertama)

    Return:
      {'LHS': [idx,...], 'LTO': [idx,...], 'RHS': [idx,...], 'RTO': [idx,...]}
    """
    results = {'LHS': [], 'LTO': [], 'RHS': [], 'RTO': []}
    label_to_name = {1: 'LHS', 2: 'LTO', 3: 'RHS', 4: 'RTO'}

    for lv, name in label_to_name.items():
        mask = (pred_labels == lv).astype(np.int32)
        if mask.sum() == 0:
            continue

        labeled, n_comp = ndimage_label(mask)
        indices = []
        for cid in range(1, n_comp + 1):
            positions = np.where(labeled == cid)[0]
            indices.append(int(positions.mean()))

        # Filter: hapus event terlalu dekat
        if len(indices) > 1:
            indices.sort()
            kept = [indices[0]]
            for idx in indices[1:]:
                if idx - kept[-1] >= min_distance:
                    kept.append(idx)
            indices = kept

        results[name] = indices

    return results

def events_to_seconds(events: dict, freq: float) -> dict:
    return {k: [round(i / freq, 4) for i in v] for k, v in events.items()}

def parse_gt_events(meta: dict) -> dict:
    """Parse ground truth dari meta.json."""
    gt = {'LHS': [], 'LTO': [], 'RHS': [], 'RTO': []}
    for pair in meta.get('leftGaitEvents', []):
        if len(pair) >= 2:
            gt['LTO'].append(int(pair[0]))
            gt['LHS'].append(int(pair[1]))
    for pair in meta.get('rightGaitEvents', []):
        if len(pair) >= 2:
            gt['RTO'].append(int(pair[0]))
            gt['RHS'].append(int(pair[1]))
    return gt

print("extract_events_from_sequence() siap.")


### CELL 2.15 — Full Trial Inference (Overlap Averaging)

In [ ]:
def predict_full_trial(model, X: np.ndarray,
                        window_size: int = WINDOW_SIZE,
                        stride: int = STRIDE,
                        num_classes: int = NUM_CLASSES,
                        device=device) -> tuple:
    """
    Inference trial penuh dengan sliding window.
    Logits pada area overlap di-rata-rata (averaging).

    Input : X (T, C) — sudah di-scaled
    Output: (pred_labels (T,), logits_full (T, num_classes))
    """
    model.eval()
    T    = X.shape[0]
    T_pad = max(T, window_size)
    pad  = T_pad - T

    if pad > 0:
        X_pad = np.vstack([X, np.zeros((pad, X.shape[1]), np.float32)])
    else:
        X_pad = X

    logits_acc = np.zeros((T_pad, num_classes), np.float64)
    count_acc  = np.zeros(T_pad, np.float64)

    starts = list(range(0, T_pad - window_size + 1, stride))
    if not starts or starts[-1] + window_size < T_pad:
        starts.append(max(0, T_pad - window_size))

    with torch.no_grad():
        for s in starts:
            e   = s + window_size
            X_w = torch.from_numpy(X_pad[s:e]).unsqueeze(0).float().to(device)
            lg  = model(X_w).squeeze(0).cpu().numpy()  # (W, C)
            logits_acc[s:e] += lg
            count_acc[s:e]  += 1.0

    count_acc  = np.where(count_acc == 0, 1, count_acc)
    logits_full = (logits_acc / count_acc[:, None])[:T]
    pred_labels = logits_full.argmax(axis=1).astype(np.int64)

    return pred_labels, logits_full

# ── Demo pada satu trial test ─────────────────────────────────
if TEST_DATA:
    _td = TEST_DATA[0]
    _preds, _logits = predict_full_trial(model, _td['X'])
    _freq  = float(_td['meta'].get('freq', 100.0))
    _evts  = extract_events_from_sequence(_preds, freq=_freq)
    _gt    = parse_gt_events(_td['meta'])

    print(f"Demo trial : {_td['trial_id']}")
    print(f"Pred shape : {_preds.shape}")
    print(f"Events pred: { {k: len(v) for k,v in _evts.items()} }")
    print(f"Events GT  : { {k: len(v) for k,v in _gt.items()} }")


### CELL 2.16 — Event-Level MAE (Millisecond)

In [ ]:
def compute_event_mae(pred_ev: dict, gt_ev: dict,
                       freq: float = 100.0) -> dict:
    """
    Hitung MAE timing error (ms) per event type.
    Gunakan greedy nearest-neighbor matching antara pred dan GT.
    MAE (ms) = mean(|t_pred - t_true|) / freq * 1000
    """
    mae = {}
    for name in ['LHS', 'LTO', 'RHS', 'RTO']:
        pred = np.array(sorted(pred_ev.get(name, [])))
        gt   = np.array(sorted(gt_ev.get(name,   [])))

        if len(pred) == 0 or len(gt) == 0:
            mae[name] = np.nan
            continue

        errors, used = [], set()
        for p in pred:
            diffs = np.abs(gt - p).copy()
            for u in used: diffs[u] = 1e9
            j = int(np.argmin(diffs))
            if diffs[j] < 1e8:
                errors.append(abs(int(p) - int(gt[j])))
                used.add(j)

        mae[name] = np.mean(errors) / freq * 1000.0 if errors else np.nan

    valid       = [v for v in mae.values() if not np.isnan(v)]
    mae['overall'] = np.mean(valid) if valid else np.nan
    return mae

print("compute_event_mae() siap.")


### CELL 2.17 — Hitung 20 Gait Parameters

17 parameter semiogram + 3 tambahan (step count, uturn count, pola uturn).
Setiap parameter yang tidak bisa dihitung → `np.nan` (tidak error).


In [ ]:
# ════════════════════════════════════════════════════════════
#  SIGNAL ANALYSIS HELPERS
# ════════════════════════════════════════════════════════════

def autocorr_norm(x: np.ndarray, max_lag: int = None) -> np.ndarray:
    """Normalized autocorrelation (lag 0 = 1.0)."""
    x  = x - x.mean()
    n  = len(x)
    ac = np.correlate(x, x, mode='full')[n-1:]
    ac = ac / (ac[0] + 1e-12)
    return ac[:max_lag+1] if max_lag is not None else ac

def harmonic_ratio(x: np.ndarray, freq: float,
                   stride_time: float, n_harmonics: int = 10) -> float:
    """
    Harmonic ratio: rasio amplitudo harmonic genap / ganjil
    dari FFT sinyal akselerasi.
    """
    try:
        if stride_time <= 0 or len(x) < 10: return np.nan
        f0     = 1.0 / stride_time
        fft_m  = np.abs(np.fft.rfft(x))
        freqs  = np.fft.rfftfreq(len(x), d=1.0/freq)
        even, odd = 0.0, 0.0
        for k in range(1, n_harmonics + 1):
            idx = int(np.argmin(np.abs(freqs - k * f0)))
            if k % 2 == 0: even += fft_m[idx]
            else:           odd  += fft_m[idx]
        return float(even / (odd + 1e-12))
    except Exception:
        return np.nan

def ldlj_acc(acc_mag: np.ndarray, freq: float, duration: float) -> float:
    """
    Log Dimensionless Jerk dari magnitude akselerasi.
    Makin kecil (kurang negatif) = gerakan lebih smooth.
    """
    try:
        if duration <= 0 or len(acc_mag) < 3: return np.nan
        jerk     = np.diff(acc_mag) * freq
        jerk_int = np.trapz(jerk**2) / freq
        pk       = np.max(np.abs(acc_mag))
        if pk <= 0 or jerk_int <= 0: return np.nan
        return float(np.log((duration**3 * pk**2) / (2 * jerk_int + 1e-12)))
    except Exception:
        return np.nan

def sparc_rotation(gyr_mag: np.ndarray, freq: float,
                   fc: float = 10.0, padlevel: int = 4) -> float:
    """
    Spectral Arc Length dari magnitude gyroscope.
    Makin besar (kurang negatif) = putaran lebih smooth.
    """
    try:
        if len(gyr_mag) < 10: return np.nan
        N_pad = int(2 ** np.ceil(np.log2(len(gyr_mag) * 2**padlevel)))
        fm    = np.abs(np.fft.rfft(gyr_mag, n=N_pad))
        fm    = fm / (fm[0] + 1e-12)
        fr    = np.fft.rfftfreq(N_pad, d=1.0/freq)
        mask  = fr <= fc
        fm, fr = fm[mask], fr[mask]
        if len(fm) < 2: return np.nan
        df   = fr[1] - fr[0]
        deriv = np.diff(fm) / (df + 1e-12)
        return float(-np.trapz(np.sqrt(1 + (deriv/fc)**2), fr[:-1]))
    except Exception:
        return np.nan

# ════════════════════════════════════════════════════════════
#  TEMPORAL HELPERS
# ════════════════════════════════════════════════════════════

def stance_swing(hs: list, to: list, freq: float) -> tuple:
    """
    Stance time = HS → TO berikutnya (kaki yang sama).
    Swing  time = TO → HS berikutnya (kaki yang sama).
    """
    hs, to = sorted(hs), sorted(to)
    stances, swings = [], []
    for h in hs:
        nxt = [t for t in to if t > h]
        if nxt: stances.append((nxt[0] - h) / freq)
    for t in to:
        nxt = [h for h in hs if h > t]
        if nxt: swings.append((nxt[0] - t) / freq)
    return np.array(stances), np.array(swings)

def double_stance(lhs, lto, rhs, rto, freq):
    """Double stance: singkat setelah setiap HS sebelum TO berikutnya."""
    all_hs = sorted(lhs + rhs)
    all_to = sorted(lto + rto)
    dst = []
    for h in all_hs:
        nxt = [t for t in all_to if t > h]
        if nxt: dst.append((nxt[0] - h) / freq)
    arr = np.array(dst)
    return arr[arr > 0] if len(arr) else np.array([])

def classify_uturn(n_steps: int) -> str:
    """
    Klasifikasi pola U-turn berdasarkan jumlah langkah.
    Rule awal — dapat diganti sesuai definisi klinis:
      Pattern I   : ≤2 langkah
      Pattern II  : 3–4 langkah
      Pattern III : >4 langkah
    """
    if   n_steps <= 2: return 'Pattern_I'
    elif n_steps <= 4: return 'Pattern_II'
    else:              return 'Pattern_III'

# ════════════════════════════════════════════════════════════
#  MAIN: COMPUTE 20 GAIT PARAMETERS
# ════════════════════════════════════════════════════════════

def compute_gait_parameters(events: dict, meta: dict,
                             df: pd.DataFrame) -> dict:
    """
    Hitung 20 gait parameters dari events + meta + raw signal DataFrame.
    Return dict — isi np.nan jika tidak bisa dihitung.
    """
    freq    = float(meta.get('freq', 100.0))
    ubounds = meta.get('uturnBoundaries', None)
    protocol= meta.get('protocol', '')
    dist    = 20.0 if '20' in protocol else (10.0 if '10' in protocol else 20.0)

    LHS = events.get('LHS', [])
    LTO = events.get('LTO', [])
    RHS = events.get('RHS', [])
    RTO = events.get('RTO', [])
    all_hs = sorted(LHS + RHS)

    # ── Temporal ─────────────────────────────────────────────
    l_str = np.diff(sorted(LHS)) / freq if len(LHS) >= 2 else np.array([])
    r_str = np.diff(sorted(RHS)) / freq if len(RHS) >= 2 else np.array([])
    all_stride = np.concatenate([l_str, r_str])

    StrT   = float(np.mean(all_stride))   if len(all_stride) else np.nan
    CVStrT = float(np.std(all_stride) / (np.mean(all_stride)+1e-12) * 100)              if len(all_stride) >= 2 else np.nan

    l_stance, l_swing = stance_swing(LHS, LTO, freq)
    r_stance, r_swing = stance_swing(RHS, RTO, freq)
    all_stance = np.concatenate([l_stance, r_stance])
    all_swing  = np.concatenate([l_swing,  r_swing])

    dst_arr = double_stance(LHS, LTO, RHS, RTO, freq)
    dstT    = float(np.mean(dst_arr)) if len(dst_arr) else np.nan
    # dstT as % of StrT
    dstT_pct = (dstT / StrT * 100)                if (not np.isnan(dstT) and not np.isnan(StrT) and StrT > 0) else np.nan
    CVdstT  = float(np.std(dst_arr) / (np.mean(dst_arr)+1e-12) * 100)               if len(dst_arr) >= 2 else np.nan
    swTr    = float(np.mean(all_swing) / (np.mean(all_stance)+1e-12))               if (len(all_swing) and len(all_stance)) else np.nan

    # Average speed
    if len(all_hs) >= 2 and not np.isnan(StrT):
        t_total = (all_hs[-1] - all_hs[0]) / freq
        V = dist / t_total if t_total > 0 else np.nan
    else:
        V = np.nan

    # Stride length
    SteL = float(V * StrT) if (not np.isnan(V) and not np.isnan(StrT)) else np.nan

    # ── U-turn ───────────────────────────────────────────────
    if ubounds and len(ubounds) == 2:
        us, ue           = int(ubounds[0]), int(ubounds[1])
        UtrT             = (ue - us) / freq
        uturn_hs         = [h for h in all_hs if us <= h <= ue]
        straight_hs      = [h for h in all_hs if not (us <= h <= ue)]
        uturn_step_count   = len(uturn_hs)
        straight_step_count= len(straight_hs)
        uturn_pattern      = classify_uturn(uturn_step_count)
    else:
        UtrT = np.nan
        uturn_step_count    = np.nan
        straight_step_count = len(all_hs)
        uturn_pattern       = 'Unknown'

    # ── Signal-based parameters ──────────────────────────────
    lb_map = get_lb_axis_map(df)
    T_sig  = df.shape[0]
    duration = T_sig / freq

    def get_col(key):
        if key in lb_map and lb_map[key] in df.columns:
            return df[lb_map[key]].fillna(0).values.astype(np.float64)
        return None

    acc_cc = get_col('CC')
    acc_ml = get_col('ML')
    acc_ap = get_col('AP')
    gyr_x  = get_col('GyrX')
    gyr_y  = get_col('GyrY')
    gyr_z  = get_col('GyrZ')

    # Magnitude acceleration
    acc_avail = [a for a in [acc_cc, acc_ml, acc_ap] if a is not None]
    if len(acc_avail) >= 3:
        n = min(len(a) for a in acc_avail)
        acc_mag = np.sqrt(sum(a[:n]**2 for a in acc_avail))
    elif acc_avail:
        acc_mag = np.abs(acc_avail[0])
    else:
        acc_mag = None

    # LDLJa
    LDLJa = ldlj_acc(acc_mag, freq, duration) if acc_mag is not None else np.nan

    # SPARCrot
    if gyr_x is not None and gyr_y is not None and gyr_z is not None:
        n = min(len(gyr_x), len(gyr_y), len(gyr_z))
        gyr_mag = np.sqrt(gyr_x[:n]**2 + gyr_y[:n]**2 + gyr_z[:n]**2)
        SPARCrot = sparc_rotation(gyr_mag, freq)
    elif gyr_x is not None:
        SPARCrot = sparc_rotation(np.abs(gyr_x), freq)
    else:
        SPARCrot = np.nan

    # RMSaML
    RMSaML = float(np.sqrt(np.mean(acc_ml**2))) if acc_ml is not None else np.nan

    # Autocorrelation peaks from CC acceleration
    if acc_cc is not None and len(acc_cc) > 10:
        max_lag = min(int(freq * 3), len(acc_cc) - 1)
        ac      = autocorr_norm(acc_cc, max_lag=max_lag)
        peaks, _ = find_peaks(ac[1:], height=0)
        P1acc   = float(ac[peaks[0]+1]) if len(peaks) >= 1 else np.nan
        P2acc   = float(ac[peaks[1]+1]) if len(peaks) >= 2 else np.nan
        P1P2acc = float(P1acc/(P2acc+1e-12))                   if (not np.isnan(P1acc) and not np.isnan(P2acc)) else np.nan
    else:
        P1acc = P2acc = P1P2acc = np.nan

    # Harmonic ratios
    st = StrT if not np.isnan(StrT) else 1.0
    iHRaAP = harmonic_ratio(acc_ap, freq, st) if acc_ap is not None else np.nan
    iHRaCC = harmonic_ratio(acc_cc, freq, st) if acc_cc is not None else np.nan
    iHRaML = harmonic_ratio(acc_ml, freq, st) if acc_ml is not None else np.nan

    return {
        'V'                  : V,
        'StrT'               : StrT,
        'UtrT'               : UtrT,
        'LDLJa'              : LDLJa,
        'SPARCrot'           : SPARCrot,
        'CVStrT'             : CVStrT,
        'CVdstT'             : CVdstT,
        'P1acc'              : P1acc,
        'P2acc'              : P2acc,
        'SteL'               : SteL,
        'RMSaML'             : RMSaML,
        'iHRaAP'             : iHRaAP,
        'iHRaCC'             : iHRaCC,
        'iHRaML'             : iHRaML,
        'P1P2acc'            : P1P2acc,
        'swTr'               : swTr,
        'dstT'               : dstT_pct,
        'straight_step_count': straight_step_count,
        'uturn_step_count'   : uturn_step_count,
        'uturn_pattern'      : uturn_pattern,
    }

print("compute_gait_parameters() siap (20 parameter).")


### CELL 2.18 — Batch Inference Semua Trial

In [ ]:
def run_batch_inference(trial_list: list, model,
                         scaler, device=device,
                         save_pred_json=True) -> list:
    """
    Full pipeline inference untuk semua trials:
      load → preprocess → predict → extract events
      → compute gait params → collect MAE
    Jika satu trial gagal, log error dan lanjutkan.
    """
    results, errors = [], []

    for info in tqdm(trial_list, desc='Batch Inference'):
        tid = info['trial_id']
        try:
            # 1. Load & preprocess
            t    = load_trial(info)
            freq = float(t['meta'].get('freq', 100.0))
            X    = preprocess_signal(t['signal'], freq=freq)
            X    = apply_scaler(X, scaler)

            # 2. Full-trial prediction
            pred_labels, _ = predict_full_trial(model, X)

            # 3. Extract predicted events
            pred_ev = extract_events_from_sequence(pred_labels, freq=freq)

            # 4. Ground truth events & MAE
            gt_ev = parse_gt_events(t['meta'])
            mae   = compute_event_mae(pred_ev, gt_ev, freq=freq)

            # 5. Gait parameters
            gait_p = compute_gait_parameters(pred_ev, t['meta'], t['df'])

            # 6. Save per-trial prediction JSON
            if save_pred_json:
                out = {
                    'trial_id'       : tid,
                    'group'          : info['group'],
                    'freq'           : freq,
                    'pred_events_idx': {k:[int(x) for x in v]
                                        for k,v in pred_ev.items()},
                    'pred_events_sec': events_to_seconds(pred_ev, freq),
                    'gt_events_idx'  : {k:[int(x) for x in v]
                                        for k,v in gt_ev.items()},
                    'gt_events_sec'  : events_to_seconds(gt_ev, freq),
                    'mae_ms'         : {k:(float(v) if not np.isnan(v) else None)
                                        for k,v in mae.items()},
                }
                with open(os.path.join(PRED_DIR, f"{tid}_pred.json"), 'w') as f:
                    json.dump(out, f, indent=2)

            # 7. Build result row
            meta = t['meta']
            row  = {
                'subject'           : meta.get('subject', info['subject_id']),
                'group'             : info['group'],
                'pathologyKey'      : meta.get('pathologyKey', ''),
                'clinicalDeficitSide': meta.get('clinicalDeficitSide', ''),
                'session'           : meta.get('session', ''),
                'trial'             : meta.get('trial', ''),
                'freq'              : freq,
                'age'               : meta.get('age', np.nan),
                'gender'            : meta.get('gender', ''),
                'height'            : meta.get('height', np.nan),
                'weight'            : meta.get('weight', np.nan),
                'BMI'               : meta.get('BMI', np.nan),
                'FMA_LE'            : meta.get('evaluationScoreValue', np.nan),
                **gait_p,
                'n_LHS'             : len(pred_ev.get('LHS',[])),
                'n_LTO'             : len(pred_ev.get('LTO',[])),
                'n_RHS'             : len(pred_ev.get('RHS',[])),
                'n_RTO'             : len(pred_ev.get('RTO',[])),
                'MAE_LHS_ms'        : mae.get('LHS', np.nan),
                'MAE_LTO_ms'        : mae.get('LTO', np.nan),
                'MAE_RHS_ms'        : mae.get('RHS', np.nan),
                'MAE_RTO_ms'        : mae.get('RTO', np.nan),
                'MAE_overall_ms'    : mae.get('overall', np.nan),
            }
            results.append(row)

        except Exception as e:
            errors.append((tid, str(e)))

    if errors:
        print(f"\n[WARN] {len(errors)} trial gagal:")
        for tid, err in errors[:10]:
            print(f"  {tid}: {err}")

    return results

print("Menjalankan batch inference untuk semua 488 trials...")
ALL_RESULTS = run_batch_inference(ALL_TRIALS, model, GLOBAL_SCALER)
print(f"Selesai — {len(ALL_RESULTS)} trials diproses.")


### CELL 2.19 — Simpan CSV dan Metrics

In [ ]:
# Buat DataFrame
df_res = pd.DataFrame(ALL_RESULTS)

# Urutan kolom
FRONT   = ['subject','group','pathologyKey','clinicalDeficitSide',
           'session','trial','freq','age','gender','height','weight','BMI','FMA_LE']
PARAMS  = ['V','StrT','UtrT','LDLJa','SPARCrot','CVStrT','CVdstT',
           'P1acc','P2acc','SteL','RMSaML','iHRaAP','iHRaCC','iHRaML',
           'P1P2acc','swTr','dstT',
           'straight_step_count','uturn_step_count','uturn_pattern']
EVENTS  = ['n_LHS','n_LTO','n_RHS','n_RTO']
MAE_C   = ['MAE_LHS_ms','MAE_LTO_ms','MAE_RHS_ms','MAE_RTO_ms','MAE_overall_ms']

ordered = [c for c in FRONT+PARAMS+EVENTS+MAE_C if c in df_res.columns]
df_res  = df_res[ordered + [c for c in df_res.columns if c not in ordered]]

df_res.to_csv(CSV_PATH, index=False)
print(f"CSV disimpan  : {CSV_PATH}")
print(f"Shape         : {df_res.shape}")

# Tambahkan event MAE ke METRICS
METRICS['event_mae'] = {
    ev: float(df_res[f'MAE_{ev}_ms'].dropna().mean())
    for ev in ['LHS','LTO','RHS','RTO']
    if f'MAE_{ev}_ms' in df_res.columns
}
METRICS['event_mae']['overall'] = float(df_res['MAE_overall_ms'].dropna().mean())                                    if 'MAE_overall_ms' in df_res.columns else np.nan

with open(METRICS_PATH, 'w') as f:
    json.dump(METRICS, f, indent=2,
              default=lambda x: None if (isinstance(x, float) and np.isnan(x)) else x)
print(f"Metrics disimpan: {METRICS_PATH}")

# Ringkasan MAE
print("\n=== Ringkasan MAE (semua trial) ===")
for col in MAE_C:
    if col in df_res.columns:
        s = df_res[col].dropna()
        print(f"  {col:20s}: mean={s.mean():.2f} ms  median={s.median():.2f} ms  n={len(s)}")


### CELL 2.20 — Visualisasi

In [ ]:
# ─── Plot 1: Training / Validation Curves ────────────────────
fig, axes = plt.subplots(1, 2, figsize=(13, 4))
ep = range(1, len(history['train_loss'])+1)
axes[0].plot(ep, history['train_loss'], label='Train', color='steelblue')
axes[0].plot(ep, history['val_loss'],   label='Val',   color='darkorange')
axes[0].set(xlabel='Epoch', ylabel='Loss', title='Loss Curve'); axes[0].legend(); axes[0].grid(alpha=.3)
axes[1].plot(ep, history['train_f1'],   label='Train', color='steelblue')
axes[1].plot(ep, history['val_f1'],     label='Val',   color='darkorange')
axes[1].set(xlabel='Epoch', ylabel='Macro F1', title='Macro F1 Curve'); axes[1].legend(); axes[1].grid(alpha=.3)
plt.tight_layout(); plt.savefig(os.path.join(OUTPUT_DIR,'training_curves.png'),dpi=150,bbox_inches='tight'); plt.show()
print("Plot 1 selesai")

# ─── Plot 2: Confusion Matrix ─────────────────────────────────
fig, ax = plt.subplots(figsize=(7,6))
cm_n = cm.astype(float) / (cm.sum(axis=1,keepdims=True)+1e-12)
sns.heatmap(cm_n, annot=True, fmt='.2f', cmap='Blues',
            xticklabels=CLASS_NAMES, yticklabels=CLASS_NAMES, ax=ax)
ax.set(xlabel='Predicted', ylabel='True', title='Confusion Matrix (Normalized) — Test')
plt.tight_layout(); plt.savefig(os.path.join(OUTPUT_DIR,'confusion_matrix.png'),dpi=150,bbox_inches='tight'); plt.show()
print("Plot 2 selesai")

# ─── Plot 3: GT vs Prediction (satu trial) ───────────────────
if TEST_TRIALS:
    _ti  = TEST_TRIALS[0]
    _t   = load_trial(_ti)
    _fq  = float(_t['meta'].get('freq',100.0))
    _Xp  = apply_scaler(preprocess_signal(_t['signal'],_fq), GLOBAL_SCALER)
    _ygt = generate_labels(_Xp.shape[0], _t['meta'])
    _ypr, _ = predict_full_trial(model, _Xp)
    _N   = min(1000, _Xp.shape[0])
    _tax = np.arange(_N)/_fq

    fig, axes = plt.subplots(3,1,figsize=(16,8),sharex=True)
    axes[0].plot(_tax, _t['signal'][:_N,0], lw=.8, color='royalblue')
    axes[0].set(ylabel='IMU Ch0', title=f"Trial: {_ti['trial_id']}"); axes[0].grid(alpha=.3)
    axes[1].step(_tax, _ygt[:_N], where='mid', color='steelblue', lw=1.5)
    axes[1].set(ylabel='Ground Truth',yticks=range(5),yticklabels=CLASS_NAMES); axes[1].grid(alpha=.3)
    axes[2].step(_tax, _ypr[:_N], where='mid', color='darkorange', lw=1.5)
    axes[2].set(ylabel='Prediction',xlabel='Time (s)',yticks=range(5),yticklabels=CLASS_NAMES); axes[2].grid(alpha=.3)
    plt.tight_layout(); plt.savefig(os.path.join(OUTPUT_DIR,'gt_vs_pred.png'),dpi=150,bbox_inches='tight'); plt.show()
    print("Plot 3 selesai")

# ─── Plot 4: MAE per event type ──────────────────────────────
mae_ev_cols = ['MAE_LHS_ms','MAE_LTO_ms','MAE_RHS_ms','MAE_RTO_ms']
if all(c in df_res.columns for c in mae_ev_cols):
    evlabels = ['LHS','LTO','RHS','RTO']
    means = [df_res[c].dropna().mean() for c in mae_ev_cols]
    stds  = [df_res[c].dropna().std()  for c in mae_ev_cols]
    colors= ['#2ecc71','#e67e22','#e74c3c','#9b59b6']
    fig, ax = plt.subplots(figsize=(7,4))
    bars = ax.bar(evlabels, means, yerr=stds, capsize=6, color=colors, alpha=.85)
    for b, m in zip(bars, means):
        ax.text(b.get_x()+b.get_width()/2, b.get_height()+1,
                f'{m:.1f}', ha='center', va='bottom', fontsize=9)
    ax.set(ylabel='MAE (ms)', title='MAE per Event Type (mean ± SD)'); ax.grid(alpha=.3, axis='y')
    plt.tight_layout(); plt.savefig(os.path.join(OUTPUT_DIR,'mae_bar.png'),dpi=150,bbox_inches='tight'); plt.show()
    print("Plot 4 selesai")

# ─── Plot 5: Histogram timing error ──────────────────────────
fig, axes = plt.subplots(2,2,figsize=(12,7))
for ax, col, lbl in zip(axes.flatten(), mae_ev_cols, evlabels):
    data = df_res[col].dropna() if col in df_res.columns else pd.Series([], dtype=float)
    ax.hist(data, bins=25, color='steelblue', edgecolor='white', alpha=.8)
    if len(data):
        ax.axvline(data.mean(), color='red', lw=1.5, ls='--',
                   label=f'Mean={data.mean():.1f}ms')
        ax.legend(fontsize=8)
    ax.set(xlabel='MAE (ms)', ylabel='Count', title=f'Timing Error — {lbl}'); ax.grid(alpha=.3)
plt.tight_layout(); plt.savefig(os.path.join(OUTPUT_DIR,'timing_error_hist.png'),dpi=150,bbox_inches='tight'); plt.show()
print("Plot 5 selesai")

# ─── Plot 6: Raw IMU contoh ───────────────────────────────────
if ALL_TRIALS:
    _raw = load_trial(ALL_TRIALS[0])
    _sig = _raw['signal'][:600]
    n_ch = min(_sig.shape[1], 6)
    fig, axes = plt.subplots(n_ch, 1, figsize=(14, 2*n_ch), sharex=True)
    if n_ch == 1: axes = [axes]
    for i, ax in enumerate(axes):
        lbl = _raw['columns'][i] if i < len(_raw['columns']) else f'Ch{i}'
        ax.plot(_sig[:,i], lw=.8, color='steelblue')
        ax.set_ylabel(lbl, fontsize=8); ax.grid(alpha=.3)
    axes[-1].set_xlabel('Sample')
    fig.suptitle(f"Raw IMU — {ALL_TRIALS[0]['trial_id']}", fontsize=11)
    plt.tight_layout(); plt.savefig(os.path.join(OUTPUT_DIR,'raw_imu.png'),dpi=150,bbox_inches='tight'); plt.show()
    print("Plot 6 selesai")

print(f"\nSemua plot disimpan ke: {OUTPUT_DIR}")


### CELL 2.21 — Notes, Limitasi, dan Next Steps

#### Output Files
| File | Keterangan |
|------|------------|
| `outputs/best_model.pt` | Model terbaik (best val Macro F1) |
| `outputs/test_metrics.json` | Frame-level metrics + event MAE |
| `outputs/event_predictions/*.json` | Predicted events per trial |
| `outputs/gait_parameters_all_trials.csv` | 20 gait parameters semua trial |
| `outputs/scaler.pkl` | Fitted StandardScaler |

#### Arsitektur
- **BiTCN**: Forward + Backward dilated TCN (dilation 2^i), residual + BN+ReLU
- **BiGRU**: 2-layer bidirectional GRU, hidden=128 → output 256-dim
- **CrossAttention**: `nn.MultiheadAttention(batch_first=True)`, Q=K=V=BiGRU output, dengan residual & LayerNorm
- Berdasarkan: Xu et al. (2025) *Bioengineering* 12(5):491

#### Limitasi
1. **EVENT_TOLERANCE** (±5 sample): dikalibrasi per dataset — coba 3–7 sample
2. **U-turn pattern** (I/II/III): rule sederhana berbasis jumlah langkah, perlu validasi klinis
3. **Harmonic ratio** tidak stabil untuk trial sangat pendek atau event sangat sedikit
4. **SteL** diestimasi dari V×StrT (bukan dari motion capture)
5. **Class imbalance** event tinggi — pertimbangkan Focal Loss untuk iterasi berikutnya

#### Next Steps
- [ ] Bandingkan dengan TCN-GRU dan BiTCN-BiGRU (ablation study)
- [ ] Analisis statistik CVA vs HS (t-test / Mann-Whitney U)
- [ ] Focal Loss untuk menangani imbalance lebih agresif
- [ ] Visualisasi semiogram polar (radar chart) per subject
- [ ] Export model ke ONNX untuk deployment
